### PR TITLE
Reduce lock contention in KubeProxy

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

**Comment by wojtek-t** [comment]
@thockin - this PR should enable your https://github.com/kubernetes/kubernetes/pull/41030 with periodic running - it should fix all your "FIXMEs" out of the box.

**Comment by wojtek-t** [comment]
@thockin @freehan - can you please take a look?

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -918,17 +894,21 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
 	}
 
 	// Figure out the new services we need to activate.
+	proxier.changeMu.Lock()
 	serviceSyncRequired, hcServices, staleServices := updateServiceMap(
 		proxier.serviceMap, &proxier.serviceChanges)
+	proxier.changeMu.Unlock()
```
Comment (freehan) [inline pkg/proxy/iptables/proxier.go]: I think it is safer to lock both updateServiceMap and updateEndpointsMap. I cannot think of any corner case on top of my head. But I feel like syncProxyRule should strictly act on a specific state in sequence. So do not process changes for endpoints after serviceMap is settled.

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -487,58 +500,38 @@ func (proxier *Proxier) SyncLoop() {
...
-	defer proxier.mu.Unlock()
-
-	change, exists := proxier.serviceChanges[namespacedName]
-	if !exists {
-		change = &serviceChange{}
-		change.previous = nil
-		proxier.serviceChanges[namespacedName] = change
-	}
-	change.current = service
+	proxier.changeMu.Lock()
```
Comment (freehan) [inline pkg/proxy/iptables/proxier.go]: move locking and update into a function? so that no need to repeat lock/unlock

**Review by freehan — COMMENTED** [review]
LGTM overall. Easy fixes.

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: Those two aren't synchronized between each other at all. I mean that it's technically possible that e.g. service is up-to-date and endpoints are lagging by 1 minute, or the other way around.
So trying to lock them together doesn't matter at all.

**Review by wojtek-t — COMMENTED** [review]
@freehan - comments applied. PTAL

**Comment by wojtek-t** [comment]
@k8s-bot unit test this

**Comment by wojtek-t** [comment]
@freehan - PTAL

**Comment by freehan** [comment]
@k8s-bot unit test this

**Comment by wojtek-t** [comment]
@freehan - this unit test failure is some regression at head - there were dozens of PRs (and also head) that is failing those since yesterday

**Pull request merged**